tikv/raftstore/snap_manager.go

```diff
@@ -0,0 +1,355 @@
+	fis, err := ioutil.ReadDir(sm.base)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	results := make([]SnapKeyWithSending, 0, len(fis))
+	for _, fi := range fis {
+		if fi.IsDir() {
+			continue
+		}
+		name := fi.Name()
+		if !strings.HasSuffix(name, metaFileSuffix) {
+			continue
+		}
+		name = name[:len(name)-len(metaFileSuffix)]
+		var key SnapKeyWithSending
+		if strings.HasPrefix(name, snapGenPrefix) {
+			key.IsSending = true
+		}
+		numberStrs := strings.Split(name, "_")
+		if len(numberStrs) != 4 {
+			return nil, errors.Errorf("failed to parse file %s", name)
+		}
+		key.SnapKey.RegionID, err = strconv.ParseUint(numberStrs[1], 10, 64)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		key.SnapKey.Term, err = strconv.ParseUint(numberStrs[2], 10, 64)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		key.SnapKey.Index, err = strconv.ParseUint(numberStrs[3], 10, 64)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		sm.registryLock.RLock()
+		_, ok := sm.registry[key.SnapKey]
+		sm.registryLock.RUnlock()
+		if ok {
+			// Skip those registered snapshot.
+			continue
+		}
+		results = append(results, key)
+	}
+	sort.Slice(results, func(i, j int) bool {
+		keyI := &results[i].SnapKey
+		keyJ := &results[j].SnapKey
+		if keyI.RegionID == keyJ.RegionID {
+			if keyI.Term == keyJ.Term {
+				if keyI.Index == keyJ.Index {
+					return !results[i].IsSending
+				}
+				return keyI.Index < keyJ.Index
+			}
+			return keyI.Term < keyJ.Term
+		}
+		return keyI.RegionID < keyJ.RegionID
+	})
+	return results, nil
+}
+
+func (sm *SnapManager) HasRegistered(key SnapKey) bool {
+	sm.registryLock.RLock()
+	_, ok := sm.registry[key]
+	sm.registryLock.RUnlock()
+	return ok
+}
+
+func (sm *SnapManager) GetTotalSnapSize() uint64 {
+	return uint64(atomic.LoadInt64(sm.snapSize))
+}
+
+func (sm *SnapManager) GetSnapshotForBuilding(key SnapKey, bundle *DBBundle) (Snapshot, error) {
+	if sm.GetTotalSnapSize() > sm.MaxTotalSize {
+		err := sm.deleteOldIdleSnaps()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return NewSnapForBuilding(sm.base, key, sm.snapSize, sm, sm.limiter)
+}
+
+func (sm *SnapManager) deleteOldIdleSnaps() error {
+	idleSnaps, err := sm.ListIdleSnap()
+	if err != nil {
+		return err
+	}
+	type snapWithModTime struct {
+		key     SnapKey
+		snap    Snapshot
+		modTime time.Time
+	}
+	snaps := make([]snapWithModTime, 0, len(idleSnaps))
+	for _, idleSnap := range idleSnaps {
+		if !idleSnap.IsSending {
+			continue
+		}
+		snap, err := sm.GetSnapshotForSending(idleSnap.SnapKey)
+		if err != nil {
+			continue
+		}
+		fi, err := snap.Meta()
+		if err != nil {
+			return err
+		}
+		snaps = append(snaps, snapWithModTime{key: idleSnap.SnapKey, snap: snap, modTime: fi.ModTime()})
+	}
+	sort.Slice(snaps, func(i, j int) bool {
+		return snaps[i].modTime.Before(snaps[j].modTime)
+	})
+	for sm.GetTotalSnapSize() > sm.MaxTotalSize {
+		if len(snaps) == 0 {
+			return errors.New("too many snapshots")
```

Comment:
Should collect several rounds?

Reply:
There maybe some old snapshots can be deleted in the next round.

Reply:
The passed duration is very short, it's unlikely that `ListIdleSnap` again can avoid this error.
And the code is simpler when we only call `ListIdleSnap` once.